File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/UserDataClient.kt
// What the settings screens ask of the user's data, which the engine owns
// (docs/architecture/user-data-engine-roadmap.md P8b). Mirrors macOS
// UserDataClient.swift.

package com.siansiansu.taigikeyboard.ime.dictionary

import com.siansiansu.taigikeyboard.engine.RustEngineBridge
import com.siansiansu.taigikeyboard.engine.backupExport
import com.siansiansu.taigikeyboard.engine.backupImport
import com.siansiansu.taigikeyboard.engine.customDictionaryDelete
import com.siansiansu.taigikeyboard.engine.customDictionaryExportCsv
import com.siansiansu.taigikeyboard.engine.customDictionaryImportCsv
import com.siansiansu.taigikeyboard.engine.customDictionaryList
import com.siansiansu.taigikeyboard.engine.customDictionarySave
import com.siansiansu.taigikeyboard.engine.customDictionarySearch
import com.siansiansu.taigikeyboard.engine.proto.BackupRefusal
import com.siansiansu.taigikeyboard.engine.proto.CustomDictionaryEntry
import com.siansiansu.taigikeyboard.engine.proto.CustomDictionaryRefusal
import com.siansiansu.taigikeyboard.engine.proto.ResetUserData
import com.siansiansu.taigikeyboard.engine.userDataReset
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.UUID

/** One word in the user's own dictionary, as the screens show and edit it. */
data class CustomDictionaryWord(
    /** Stable across edits: the engine stores an edit under the same id. */
    val id: String = UUID.randomUUID().toString(),
    val roman: String,
    val hanzi: String,
)

data class CustomDictionaryImportResult(
    val imported: Int,
    val skipped: Int,
)

/** Rows a `.taigi` restore merged, per store. */
data class BackupImportResult(
    val customDict: Int,
    val frequency: Int,
    val association: Int,
)

/** Why a user-data request did nothing. */
sealed class UserDataException(
    message: String,
) : Exception(message) {
    /** The round-trip failed or the engine refused the request; the bridge logged why. */
    class EngineUnavailable(
        op: String,
    ) : UserDataException("the engine did not answer $op")

    /** Something the user can be told — a full dictionary, an unusable file — in the engine's words. */
    class Refused(
        val refusal: CustomDictionaryRefusal,
        detail: String,
    ) : UserDataException(detail)

    /** Not a `.taigi` file, or one from before version 1. */
    class BackupRefused(
        val refusal: BackupRefusal,
    ) : UserDataException("backup refused: $refusal")

    /** A reset that emptied some stores and not others, one line per store that failed. */
    class NotEmptied(
        val failures: List<String>,
    ) : UserDataException(failures.joinToString("\n"))
}

/**
 * The user-data requests the settings screens make. Every call is an engine
 * round-trip that may wait on SQLite — and right after launch on the engine
 * finishing its takeover of the old files — so the shipped client runs each
 * on `Dispatchers.IO`. An interface so view models can be driven from JVM
 * tests, which cannot load the engine.
 */
interface UserDataClient {
    /** Every word, newest edit first. */
    suspend fun listAll(): List<CustomDictionaryWord>

    suspend fun save(word: CustomDictionaryWord)

    suspend fun delete(id: String)

    /** Empties the custom dictionary. */
    suspend fun deleteAll()

    suspend fun exportCsv(): ByteArray

    /** A `roman,hanzi` CSV file's bytes, at most [MAX_IMPORT_FILE_BYTES]. */
    suspend fun importCsv(csv: ByteArray): CustomDictionaryImportResult

    /** Empties what the keyboard learned — counts, bigrams, learned phrases — and leaves the custom dictionary alone. */
    suspend fun clearLearningRecords()

    /**
     * The dictionary search's lookup, by the key [query] derives under the
     * settings [inputMode] (`poj` / `tl` / `tps`); empty when nothing matches
     * or the engine did not answer.
     */
    suspend fun search(
        query: String,
        inputMode: String,
        limit: Int,
    ): List<CustomDictionaryWord>

    suspend fun exportBackup(appVersion: String): ByteArray

    suspend fun importBackup(backup: ByteArray): BackupImportResult

    companion object {
        /**
         * The largest file an import reads. CROSS-PLATFORM INVARIANT — the
         * engine refuses the same size (`engine/userdata/src/csv.rs`); checked
         * on this side too so a huge file is refused before it is read.
         */
        const val MAX_IMPORT_FILE_BYTES: Long = 5L * 1024 * 1024
    }
}

/** The shipped client: the engine's user-data ops, each on `Dispatchers.IO`. */
object EngineUserDataClient : UserDataClient {
    override suspend fun listAll(): List<CustomDictionaryWord> =
        engine("customDictionaryList") { RustEngineBridge.customDictionaryList(filter = "", limit = 0, offset = 0) }
            .entriesList
            .map(::word)

    override suspend fun save(word: CustomDictionaryWord) {
        val saved = engine("customDictionarySave") { RustEngineBridge.customDictionarySave(word.id, word.roman, word.hanzi) }
        if (saved.refusal != CustomDictionaryRefusal.CUSTOM_DICTIONARY_REFUSAL_NONE) {
            throw UserDataException.Refused(saved.refusal, saved.detail)
        }
    }

    override suspend fun delete(id: String) {
        engine("customDictionaryDelete") { RustEngineBridge.customDictionaryDelete(id) }
    }

    override suspend fun deleteAll() = reset(ResetUserData.newBuilder().setCustomDictionary(true).build())

    override suspend fun exportCsv(): ByteArray = engine("customDictionaryExportCsv") { RustEngineBridge.customDictionaryExportCsv() }

    override suspend fun importCsv(csv: ByteArray): CustomDictionaryImportResult {
        val imported = engine("customDictionaryImportCsv") { RustEngineBridge.customDictionaryImportCsv(csv) }
        if (imported.refusal != CustomDictionaryRefusal.CUSTOM_DICTIONARY_REFUSAL_NONE) {
            throw UserDataException.Refused(imported.refusal, imported.detail)
        }
        return CustomDictionaryImportResult(imported = imported.imported, skipped = imported.skipped)
    }

    override suspend fun clearLearningRecords() =
        reset(
            ResetUserData
                .newBuilder()
                .setFrequency(true)
                .setAssociation(true)
                .setLearnedPhrases(true)
                .build(),
        )

    override suspend fun search(
        query: String,
        inputMode: String,
        limit: Int,
    ): List<CustomDictionaryWord> =
        withContext(Dispatchers.IO) {
            RustEngineBridge.customDictionarySearch(query, inputMode, limit)?.map(::word) ?: emptyList()
        }

    override suspend fun exportBackup(appVersion: String): ByteArray = engine("backupExport") { RustEngineBridge.backupExport(appVersion) }

    override suspend fun importBackup(backup: ByteArray): BackupImportResult {
        val imported = engine("backupImport") { RustEngineBridge.backupImport(backup) }
        if (imported.refusal != BackupRefusal.BACKUP_REFUSAL_NONE) {
            throw UserDataException.BackupRefused(imported.refusal)
        }
        return BackupImportResult(
            customDict = imported.customDictionary,
            frequency = imported.frequency,
            association = imported.association,
        )
    }

    /** Empties the stores [request] selects; every one is attempted, and the ones that could not be emptied are reported together. */
    private suspend fun reset(request: ResetUserData) {
        val removed = engine("userDataReset") { RustEngineBridge.userDataReset(request) }
        if (removed.failuresCount > 0) throw UserDataException.NotEmptied(removed.failuresList)
    }

    /** One request on IO; `null` (a failed or refused round-trip, logged by the bridge) throws. */
    private suspend fun <T : Any> engine(
        op: String,
        request: () -> T?,
    ): T = withContext(Dispatchers.IO) { request() } ?: throw UserDataException.EngineUnavailable(op)

    private fun word(entry: CustomDictionaryEntry) = CustomDictionaryWord(id = entry.id, roman = entry.roman, hanzi = entry.hanzi)
}

File: ios/Sources/TaigiKeyboard/Lexicon/Services/UserDataClient.swift
// What the app's Dictionary pages ask of the user's data, which the engine
// owns (docs/architecture/user-data-engine-roadmap.md P7b). Mirrors macOS
// `UserDataClient.swift` and Android `UserDataClient.kt`.

import Foundation

/// Rows a `.taigi` restore merged, per store.
struct BackupImportResult: Equatable {
    let customDict: Int
    let frequency: Int
    let association: Int
}

/// A custom-dictionary import the user can be told about. Engine-layer error —
/// a plain typed enum, unaware of the App/Strings presentation layer; the App
/// layer maps each case to a localized message at the display boundary
/// (`CustomDictionaryView.localizedImportMessage`), mirroring Android where
/// `CustomDictionaryScreen` resolves the refusal. `errorDescription` carries an
/// English developer fallback so an unmapped case stays diagnosable.
enum CustomDictionaryError: LocalizedError {
    case invalidCSVData
    case invalidCSVFormat
    case fileTooLarge
    case tooManyEntries

    var errorDescription: String? {
        switch self {
        case .invalidCSVData: "Invalid CSV data"
        case .invalidCSVFormat: "Invalid CSV format"
        case .fileTooLarge: "File too large"
        case .tooManyEntries: "Too many entries"
        }
    }
}

/// A `.taigi` file the engine would not restore.
enum BackupError: LocalizedError {
    case unreadable
    case unsupportedVersion

    var errorDescription: String? {
        switch self {
        case .unreadable: "Not a readable backup file"
        case .unsupportedVersion: "Unsupported backup version"
        }
    }
}

/// The round-trip failed or the engine refused the request (before the open);
/// the bridge recorded why in its diagnostics.
struct UserDataUnavailable: LocalizedError {
    let op: String

    var errorDescription: String? {
        "The engine did not answer \(op)"
    }
}

/// The engine would not do what was asked — an unsearchable romanization, a
/// store a reset could not empty — in its own words.
struct UserDataRefused: LocalizedError {
    let detail: String

    var errorDescription: String? {
        detail
    }
}

/// The user-data requests the app's pages make. Each is an engine round-trip
/// that may wait on SQLite — and right after launch on the engine finishing
/// its takeover of the old files — so the shipped client runs every one off
/// the main actor. A protocol so view models can be driven from tests, which
/// never open the user data (the test process shares the app's engine).
protocol UserDataClient: Sendable {
    /// Every word, newest edit first.
    func listAll() async throws -> [CustomDictionaryEntry]
    /// A refusal (a full dictionary, an unsearchable romanization) throws.
    func save(_ entry: CustomDictionaryEntry) async throws
    func delete(id: String) async throws
    /// Empties the custom dictionary.
    func deleteAll() async throws
    func exportCSV() async throws -> Data
    /// A `roman,hanzi` CSV file the user picked; a refusal throws `CustomDictionaryError`.
    func importCSV(url: URL) async throws -> (imported: Int, skipped: Int)
    /// Empties what the keyboard learned — counts, bigrams, learned phrases —
    /// and leaves the custom dictionary alone.
    func clearLearningRecords() async throws
    /// The dictionary search's lookup, by the key `query` derives under the
    /// settings `inputMode`; empty when nothing matches or the engine did not answer.
    func search(query: String, mode: InputMode, limit: Int) async -> [CustomDictionaryEntry]
    func exportBackup(appVersion: String) async throws -> Data
    /// A `.taigi` file the user picked; a refusal throws `BackupError`.
    func importBackup(url: URL) async throws -> BackupImportResult
}

/// The shipped client: the engine's user-data ops, one at a time on a queue
/// of their own.
struct EngineUserDataClient: UserDataClient {
    /// The page requests, off the main actor. A GCD queue rather than a task
    /// per request: a request can wait on SQLite — right after launch on the
    /// engine finishing its takeover — and that wait must not hold a thread of
    /// the shared pool.
    private static let requests = DispatchQueue(label: "EngineUserDataClient.requests", qos: .userInitiated)

    /// The largest file an import reads. CROSS-PLATFORM INVARIANT — the engine
    /// refuses the same size (`engine/userdata/src/csv.rs`); checked on this
    /// side too so a huge file is refused before it is read into memory.
    static let maxImportFileBytes = 5 * 1024 * 1024

    func listAll() async throws -> [CustomDictionaryEntry] {
        try await engine("customDictionaryList") { RustEngineBridge.customDictionaryList() }
            .entries
            .map(CustomDictionaryEntry.init)
    }

    func save(_ entry: CustomDictionaryEntry) async throws {
        let saved = try await engine("customDictionarySave") {
            RustEngineBridge.customDictionarySave(id: entry.id, roman: entry.roman, hanzi: entry.hanzi)
        }
        if saved.refusal == .full {
            throw CustomDictionaryError.tooManyEntries
        }
        if saved.refusal != .none {
            throw UserDataRefused(detail: saved.detail)
        }
    }

    func delete(id: String) async throws {
        _ = try await engine("customDictionaryDelete") { RustEngineBridge.customDictionaryDelete(id: id) }
    }

    func deleteAll() async throws {
        var reset = Taigi_Engine_ResetUserData()
        reset.customDictionary = true
        try await self.reset(reset)
    }

    func exportCSV() async throws -> Data {
        try await engine("customDictionaryExportCSV") { RustEngineBridge.customDictionaryExportCSV() }
    }

    func importCSV(url: URL) async throws -> (imported: Int, skipped: Int) {
        let csv = try await Self.read(url, sizeLimit: Self.maxImportFileBytes)
        let imported = try await engine("customDictionaryImportCSV") { RustEngineBridge.customDictionaryImportCSV(csv) }
        switch imported.refusal {
        case .none: return (imported: Int(imported.imported), skipped: Int(imported.skipped))
        case .fileTooLarge: throw CustomDictionaryError.fileTooLarge
        case .full: throw CustomDictionaryError.tooManyEntries
        case .notUtf8: throw CustomDictionaryError.invalidCSVData
        default: throw CustomDictionaryError.invalidCSVFormat
        }
    }

    func clearLearningRecords() async throws {
        var reset = Taigi_Engine_ResetUserData()
        reset.frequency = true
        reset.association = true
        reset.learnedPhrases = true
        try await self.reset(reset)
    }

    func search(query: String, mode: InputMode, limit: Int) async -> [CustomDictionaryEntry] {
        let found = try? await engine("customDictionarySearch") {
            RustEngineBridge.customDictionarySearch(query: query, inputMode: mode.rawValue, limit: limit)
        }
        return found?.map(CustomDictionaryEntry.init) ?? []
    }

    func exportBackup(appVersion: String) async throws -> Data {
        try await engine("backupExport") { RustEngineBridge.backupExport(appVersion: appVersion) }
    }

    func importBackup(url: URL) async throws -> BackupImportResult {
        let backup = try await Self.read(url, sizeLimit: nil)
        let imported = try await engine("backupImport") { RustEngineBridge.backupImport(backup) }
        switch imported.refusal {
        case .none:
            return BackupImportResult(
                customDict: Int(imported.customDictionary),
                frequency: Int(imported.frequency),
                association: Int(imported.association),
            )
        case .unsupportedVersion: throw BackupError.unsupportedVersion
        default: throw BackupError.unreadable
        }
    }

    /// Empties the stores `request` selects; every one is attempted, and the
    /// ones that could not be emptied are reported together.
    private func reset(_ request: Taigi_Engine_ResetUserData) async throws {
        let removed = try await engine("userDataReset") { RustEngineBridge.userDataReset(request) }
        if !removed.failures.isEmpty {
            throw UserDataRefused(detail: removed.failures.joined(separator: "\n"))
        }
    }

    /// A file the user picked, read off the main actor under its
    /// security scope — refused before the read when over `sizeLimit`.
    private static func read(_ url: URL, sizeLimit: Int?) async throws -> Data {
        try await withCheckedThrowingContinuation { continuation in
            requests.async {
                continuation.resume(with: Result {
                    let accessing = url.startAccessingSecurityScopedResource()
                    defer {
                        if accessing {
                            url.stopAccessingSecurityScopedResource()
                        }
                    }
                    if let sizeLimit {
                        let size = try url.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? 0
                        if size > sizeLimit {
                            throw CustomDictionaryError.fileTooLarge
                        }
                    }
                    return try Data(contentsOf: url)
                })
            }
        }
    }

    /// One request on `requests`; `nil` (a failed or refused round-trip,
    /// recorded by the bridge) throws.
    private func engine<T: Sendable>(_ op: String, _ request: @escaping @Sendable () -> T?) async throws -> T {
        let answer = await withCheckedContinuation { continuation in
            Self.requests.async { continuation.resume(returning: request()) }
        }
        guard let answer else {
            throw UserDataUnavailable(op: op)
        }
        return answer
    }
}

extension CustomDictionaryEntry {
    init(_ entry: Taigi_Engine_CustomDictionaryEntry) {
        self.init(id: entry.id, roman: entry.roman, hanzi: entry.hanzi)
    }
}
